### PR TITLE
cleanup(libsinsp,libscap,libpman): cleanups, fixes for return values, memory management, allocations

### PR DESCRIPTION
**Pull request by LucaGuerra**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug
/kind cleanup

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area libscap

/area libpman

/area libsinsp

**Does this PR require a change in the driver versions?**

No

**What this PR does / why we need it**:

This PR cumulates a number of stability fixes to check the following:
* Return values
* Memory allocation functions
* Potential invalid pointer access

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
cleanup(libsinsp,libscap,libpman): cleanups, fixes for return values, memory management, allocations
```

### CODE REVIEW

**Comment by LucaGuerra** [comment]
/hold

accidentally a commit

File: userspace/libscap/engine/gvisor/scap_gvisor.cpp [hunk 1]
```diff
@@ -65,14 +65,17 @@ int32_t sandbox_entry::expand_buffer(size_t size)
...
+		}
 	} else
 	{
 		new_buf = realloc(m_buf.buf, size);
-	}
-
-	if (new_buf == nullptr)
-	{
-		return SCAP_FAILURE;
+		if (new_buf == nullptr)
```
Comment (FedeDP) [inline userspace/libscap/engine/gvisor/scap_gvisor.cpp:75]: Can we wrap this check just once after the `if/else` statement?

File: userspace/libsinsp/filterchecks.cpp [hunk 2]
```diff
@@ -80,6 +80,10 @@ int32_t gmt2local(time_t t)
 	gmt = &sgmt;
 	*gmt = *gmtime(&t);
 	loc = localtime(&t);
+	if(loc == NULL)
+	{
+		throw sinsp_exception("cannot get localtime");
```
Comment (FedeDP) [inline userspace/libsinsp/filterchecks.cpp:85]: Can we have a proper fallback here?
Like, just return 0, i don't, and manage the issue in the `sinsp_utils::ts_to_string`.
NOTE: a line above we are de-referencing a possible NULL ptr:
```
*gmtime(&t);
```
Please fix it too.

File: userspace/libsinsp/filterchecks.cpp [hunk 3]
```diff
@@ -5235,7 +5239,12 @@ uint8_t* sinsp_filter_check_tracer::extract_args(sinsp_partial_tracer* pae, OUT
 
 	if(m_storage_size < encoded_args_len)
 	{
-		m_storage = (char*)realloc(m_storage, encoded_args_len);
+		char *reduced_storage = (char*)realloc(m_storage, encoded_args_len);
+		if(reduced_storage == NULL)
+		{
+			return NULL;
```
Comment (FedeDP) [inline userspace/libsinsp/filterchecks.cpp:5245]: Not freeing `m_storage` here?
But we are freeing it right below (in `extract_tracer`).

File: userspace/libsinsp/tracers.cpp [hunk 4]
```diff
@@ -40,12 +40,12 @@ sinsp_tracerparser::~sinsp_tracerparser()
 
 void sinsp_tracerparser::set_storage_size(uint32_t newsize)
 {
-	m_storage = (char*)realloc(m_storage, newsize);
-	if(m_storage == NULL)
+	char *tmp_storage = (char*)realloc(m_storage, newsize);
+	if(tmp_storage == NULL)
 	{
 		throw sinsp_exception("memory allocation error in sinsp_tracerparser::process_event_data.");
```
Comment (FedeDP) [inline userspace/libsinsp/tracers.cpp:46]: Not freeing `m_storage`?

File: userspace/libscap/engine/gvisor/scap_gvisor.cpp (hunk 1 above)
Comment (therealbobo) [reply userspace/libscap/engine/gvisor/scap_gvisor.cpp:75]: I think so: it was done this way due to a wrong free in the second if 😓

Comment (LucaGuerra) [reply userspace/libscap/engine/gvisor/scap_gvisor.cpp:75]: heh, good point

File: userspace/libsinsp/filterchecks.cpp (hunk 3 above)
Comment (therealbobo) [reply userspace/libsinsp/filterchecks.cpp:5245]: `m_storage` is freed by the destructor 🤔

File: userspace/libsinsp/filterchecks.cpp [hunk 5]
```diff
@@ -5874,7 +5888,13 @@ inline uint8_t* sinsp_filter_check_evtin::extract_tracer(sinsp_evt *evt, sinsp_p
 
 		if(m_storage_size < encoded_tags_len)
 		{
-			m_storage = (char*)realloc(m_storage, encoded_tags_len);
+			char *reduced_storage = (char*)realloc(m_storage, encoded_tags_len);
+			if(reduced_storage == NULL)
+			{
+				free(m_storage);
```
Comment (therealbobo) [inline userspace/libsinsp/filterchecks.cpp:5894]: I think this is a double free on `m_storage`.

File: userspace/libsinsp/filterchecks.cpp [hunk 6]
```diff
@@ -5935,7 +5955,13 @@ inline uint8_t* sinsp_filter_check_evtin::extract_tracer(sinsp_evt *evt, sinsp_p
 
 		if(m_storage_size < encoded_args_len)
 		{
-			m_storage = (char*)realloc(m_storage, encoded_args_len);
+			char *reduced_storage = (char*)realloc(m_storage, encoded_args_len);
+			if(reduced_storage == NULL)
+			{
+				free(m_storage);
```
Comment (therealbobo) [inline userspace/libsinsp/filterchecks.cpp:5961]: Here another double free on `m_storage`.

File: userspace/libsinsp/filterchecks.cpp (hunk 3 above)
Comment (LucaGuerra) [reply userspace/libsinsp/filterchecks.cpp:5245]: Thanks for checking, I believe we don't need to free `m_storage` in two instances below

File: userspace/libsinsp/filterchecks.cpp (hunk 2 above)
Comment (LucaGuerra) [reply userspace/libsinsp/filterchecks.cpp:90]: Looks like this is only used to check with `(0)`, on failure we probably have bigger problems.

Re. gmtime() call, done!

File: userspace/libsinsp/parsers.cpp [hunk 7]
```diff
@@ -4908,6 +4931,10 @@ void sinsp_parser::parse_select_poll_epollwait_enter(sinsp_evt *evt)
 	if(evt->m_tinfo->m_lastevent_data == NULL)
 	{
 		evt->m_tinfo->m_lastevent_data = reserve_event_buffer();
+		if(evt->m_tinfo->m_lastevent_data == NULL)
```
Comment (jasondellaluce) [inline userspace/libsinsp/parsers.cpp:4934]: I think this should be handled like above (ASSERT, log, and return). https://github.com/falcosecurity/libs/pull/888/files#diff-b1e2f413bae0df2c1d50354a44678c2350f5ce108a3609fe14896be2529873f8R889.

File: userspace/libsinsp/threadinfo.cpp [hunk 8]
```diff
@@ -772,7 +776,12 @@ sinsp_threadinfo* sinsp_threadinfo::get_parent_thread()
 
 sinsp_fdinfo_t* sinsp_threadinfo::add_fd(int64_t fd, sinsp_fdinfo_t *fdinfo)
 {
-	sinsp_fdinfo_t* res = get_fd_table()->add(fd, fdinfo);
+	sinsp_fdtable* fd_table_ptr = get_fd_table();
+	if(fd_table_ptr == NULL)
+	{
+		return NULL;
```
Comment (jasondellaluce) [inline userspace/libsinsp/threadinfo.cpp:782]:
```suggestion
		ASSERT(false);
		return NULL;
```

File: userspace/libsinsp/threadinfo.cpp [hunk 9]
```diff
@@ -784,14 +793,23 @@ sinsp_fdinfo_t* sinsp_threadinfo::add_fd(int64_t fd, sinsp_fdinfo_t *fdinfo)
 
 void sinsp_threadinfo::remove_fd(int64_t fd)
 {
-	get_fd_table()->erase(fd);
+	sinsp_fdtable* fd_table_ptr = get_fd_table();
+	if(fd_table_ptr == NULL)
+	{
+		return;
```
Comment (jasondellaluce) [inline userspace/libsinsp/threadinfo.cpp:799]:
```suggestion
		ASSERT(false);
		return;
```

File: userspace/libsinsp/threadinfo.cpp [hunk 10]
```diff
@@ -784,14 +793,23 @@ sinsp_fdinfo_t* sinsp_threadinfo::add_fd(int64_t fd, sinsp_fdinfo_t *fdinfo)
...
 }
 
 bool sinsp_threadinfo::is_bound_to_port(uint16_t number)
 {
 	std::unordered_map<int64_t, sinsp_fdinfo_t>::iterator it;
 
 	sinsp_fdtable* fdt = get_fd_table();
+	if(fdt == NULL)
+	{
+		return false;
```
Comment (jasondellaluce) [inline userspace/libsinsp/threadinfo.cpp:811]:
```suggestion
		ASSERT(false);
		return false;
```

File: userspace/libsinsp/threadinfo.cpp [hunk 11]
```diff
@@ -819,6 +837,10 @@ bool sinsp_threadinfo::uses_client_port(uint16_t number)
...
```
Comment (jasondellaluce) [inline userspace/libsinsp/threadinfo.cpp:842]:
```suggestion
		ASSERT(false);
		return false;
```

File: userspace/libsinsp/threadinfo.cpp [hunk 12]
```diff
@@ -1475,7 +1509,13 @@ void sinsp_thread_manager::remove_thread(int64_t tid, bool force)
 		//
 		if((tinfo->m_pid == tinfo->m_tid) || tinfo->m_flags & PPM_CL_IS_MAIN_THREAD)
 		{
-			std::unordered_map<int64_t, sinsp_fdinfo_t>* fdtable = &(tinfo->get_fd_table()->m_table);
+			sinsp_fdtable* fd_table_ptr = tinfo->get_fd_table();
+			if(fd_table_ptr == NULL)
+			{
+				return;
```
Comment (jasondellaluce) [inline userspace/libsinsp/threadinfo.cpp:1515]:
```suggestion
				ASSERT(false);
				return;
```

File: userspace/libsinsp/threadinfo.cpp [hunk 13]
```diff
@@ -1586,7 +1626,12 @@ void sinsp_thread_manager::update_statistics()
 	m_inspector->m_stats.m_n_fds = 0;
 	for(threadinfo_map_iterator_t it = m_threadtable.begin(); it != m_threadtable.end(); it++)
 	{
-		m_inspector->m_stats.m_n_fds += it->second.get_fd_table()->size();
+		sinsp_fdtable* fd_table_ptr = it->second.get_fd_table();
+		if(fd_table_ptr == NULL)
+		{
+			return;
```
Comment (jasondellaluce) [inline userspace/libsinsp/threadinfo.cpp:1632]:
```suggestion
			ASSERT(false);
			return;
```

**Review by jasondellaluce — COMMENTED** [review]
Left some minor suggestions! Superb work!

File: userspace/libsinsp/parsers.cpp (hunk 7 above)
Comment (LucaGuerra) [reply userspace/libsinsp/parsers.cpp:4934]: let me double check these two cases

**Comment by LucaGuerra** [comment]
@jasondellaluce thanks for the comments, extra assertions wouldn't hurt, going to accept them

File: userspace/libsinsp/filterchecks.cpp (hunk 5 above)
Comment (LucaGuerra) [reply userspace/libsinsp/filterchecks.cpp]: done! https://github.com/falcosecurity/libs/pull/888/commits/e90f4f3d3e86e07721d8ea5d03d291bb26c8fbfd

File: userspace/libsinsp/filterchecks.cpp (hunk 6 above)
Comment (LucaGuerra) [reply userspace/libsinsp/filterchecks.cpp]: done! https://github.com/falcosecurity/libs/pull/888/commits/e90f4f3d3e86e07721d8ea5d03d291bb26c8fbfd

File: userspace/chisel/chisel_api.cpp [hunk 14]
```diff
@@ -772,6 +772,10 @@ int lua_cbacks::get_thread_table_int(lua_State *ls, bool include_fds, bool bareb
 		// If not, skip this thread
 		//
 		sinsp_fdtable* fdtable = tinfo.get_fd_table();
+		if(fdtable == nullptr)
+		{
+			throw sinsp_exception("cannot get fd table in lua_cbacks::get_thread_table_int");
```
[the remark this answers is not in the document]
Comment (LucaGuerra) [reply userspace/chisel/chisel_api.cpp]: makes sense, added a comment

File: userspace/libsinsp/parsers.cpp (hunk 7 above)
Comment (LucaGuerra) [reply userspace/libsinsp/parsers.cpp:4933]: ok this is the same as a `malloc()` failure so we can keep as a `throw()` for consistency, like other ones in the file. Thanks a lot for checking!

**Review by jasondellaluce — APPROVED** [review]

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**